src/collector/top_score_collector.rs

```diff
@@ -736,6 +722,64 @@ impl SegmentCollector for TopScoreSegmentCollector {
     }
 }
 
+pub struct TopNComputer<Score, DocId> {
+    buffer: Vec<ComparableDoc<Score, DocId>>,
+    max_size: usize,
+    pub threshold: Option<Score>,
+}
+
+impl<Score, DocId> TopNComputer<Score, DocId>
+where
+    Score: PartialOrd + Clone,
+    DocId: Ord + Clone,
+{
+    pub fn new(top_n: usize) -> Self {
+        let max_size = top_n.max(1) * 2;
+        TopNComputer {
+            buffer: Vec::with_capacity(max_size),
+            max_size,
+            threshold: None,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn push(&mut self, doc: ComparableDoc<Score, DocId>) {
+        if let Some(last_median) = self.threshold.clone() {
+            if doc.feature < last_median {
+                return;
+            }
+        }
+        // Check on capacity should eleminate the check in `push`
+        if self.buffer.len() == self.buffer.capacity() {
+            let median = self.truncate_median();
+            self.threshold = Some(median);
+        }
+
+        self.buffer.push(doc);
+    }
+
+    #[inline(never)]
+    fn truncate_median(&mut self) -> Score {
+        let len = self.buffer.len();
+        // Use select_nth_unstable to find the median score
+        let (_, median_el, _) = self.buffer.select_nth_unstable(len / 2);
+
+        let median_score = median_el.feature.clone();
+        // Remove all elements below the median
+        self.buffer.truncate(len / 2);
+
+        median_score
+    }
+
+    pub(crate) fn into_iter_sorted(self) -> impl Iterator<Item = ComparableDoc<Score, DocId>> {
+        let mut sorted_buffer = self.buffer;
+        sorted_buffer.sort_unstable();
+
+        // Return the sorted top N elements
+        sorted_buffer.into_iter().take(self.max_size / 2)
```

Comment:
Couldn't this use `sorted_buffer.capacity()` and thereby avoid storing `max_size` at all?

I also wonder if truncation before calling `.into_iter()` isn't preferable to usage of `.take(..)` as it simplifies the returned iterator, including returning `impl ExactSizeIterator<Item = ...>`.

Reply:
`Vec::with_capacity` contract is
> The vector will be able to hold at least capacity elements without reallocating.

Current implementation seems to always give exactly the requested capacity (except for ZST), but it could in theory give more, in which case we could return more docs than initially requested

Reply:
I think if the implementation is to be robust against that, the check line 753 should also use `max_size` instead of `capacity`, shouldn't it?

(It would lead to more than `2K` documents being tracked and hence the first `K` could be out of order w.r.t. the median at `len() / 2 == capacity() / 2 > K`? Or does sorting the whole buffer at the end take care of that?)

(Wouldn't it actually suffice to call `truncate_median` in `into_sorted_iter` and then sort only the remaining `K` elements generally (thereby also avoid the `take` adaptor)?)

Reply:
Having the `truncate_median` logic work on a little bit more capacity is no issue.

The check in line 753 did not work, so I replaced it with some `unsafe`

> (Wouldn't it actually suffice to call truncate_median in into_sorted_iter and then sort only the remaining K elements generally (thereby also avoiding the take adaptor)?)

Yes, but I don't think the performance is worth it the lower readability

Reply:
> Yes, but I don't think the performance is worth it the lower readability

To be honest, I am not sure that readability actually suffers:

```rust
pub(crate) fn into_iter_sorted(self) -> impl ExactSizeIterator<Item = ComparableDoc<Score, DocId>> {
    self.truncate_median();
    self.buffer.sort_unstable();
    self.buffer.into_iter()
}
```

Reply:
anyone proofreading would need to check `truncate_median` truncates exact to top n ... which it doesn't

Reply:
> which it doesn't

If capacity is over-allocated, but since you went for `spare_capacity` anyway, you could actually enforce that by checking `self.buffer.len() == self.max_size` instead of `self.buffer.capacity()`.

I agree that this entangles this function with the invariants of `TopNComputer`, but I am not sure that `TopNCompuer` can be reviewed piece by piece in any case as the contents of `self.buffer` is really only meaningful here due to these invariants. 

That said, I would still suggest doing at least

```rust
pub(crate) fn into_iter_sorted(self) -> impl ExactSizeIterator<Item = ComparableDoc<Score, DocId>> {
        let mut sorted_buffer = self.buffer;
        sorted_buffer.sort_unstable();

        // Return the sorted top N elements
        sorted_buffer.truncate(self.top_n)
        sorted_buffer.into_iter()
}
```

to simplify the returned iterator.

Reply:
I don't think that simplifies the code or the returned Iterator

Reply:
It definitely avoids "injecting" the code for the `Take` adaptor into the caller, c.f. https://doc.rust-lang.org/stable/src/core/iter/adapters/take.rs.html#35